File: 167. Two Sum II - Input Array Is Sorted/main.cpp
#include <vector>

using std::lower_bound;
using std::vector;

class Solution {
public:
    vector<int> twoSum(vector<int> &numbers, int target) {
        int left = 0;
        int right = numbers.size() - 1;

        while (left < right) {
            int sum = numbers[left] + numbers[right];
            if (sum < target) {
                left++;
            } else if (sum > target) {
                right--;
            } else {
                break;
            }
        }

        return {left + 1, right + 1};
    }
};

class Solution {
public:
    vector<int> twoSum(vector<int> &numbers, int target) {
        for (int i = 0; i < numbers.size(); i++) {
            int object = target - numbers[i];
            auto it = lower_bound(numbers.begin() + i + 1, numbers.end(), object);
            if (it != numbers.end() && *it == object) {
                return {i + 1, (int)(it - numbers.begin()) + 1};
            }
        }

        return {-1, -1};
    }
};

File: 167. Two Sum II - Input Array Is Sorted/main.rs
struct Solution;

impl Solution {
    pub fn two_sum(numbers: Vec<i32>, target: i32) -> Vec<i32> {
        let mut left = 0;
        let mut right = numbers.len() as i32 - 1;

        while left < right {
            let sum = numbers[left as usize] + numbers[right as usize];
            if sum < target {
                left += 1;
            } else if sum > target {
                right -= 1;
            } else {
                break;
            }
        }

        vec![left + 1, right + 1]
    }
}

impl Solution {
    pub fn two_sum(numbers: Vec<i32>, target: i32) -> Vec<i32> {
        for first in 0..numbers.len() {
            let object = target - numbers[first];
            if let Ok(index) = numbers[first + 1..].binary_search(&object) {
                return vec![first as i32 + 1, index as i32 + first as i32 + 2];
            }
        }

        vec![-1, -1]
    }
}
